proxy/state.go

```diff
@@ -0,0 +1,124 @@
+	refreshCooldown time.Duration
+	sessionStore    sessions.SessionStore
+	sessionLoaders  []sessions.SessionLoader
+	jwtClaimHeaders []string
+	authzClient     envoy_service_auth_v2.AuthorizationClient
+}
+
+func newProxyStateFromConfig(cfg *config.Config) (*proxyState, error) {
+	err := ValidateOptions(cfg.Options)
+	if err != nil {
+		return nil, err
+	}
+
+	state := new(proxyState)
+	state.sharedKey = cfg.Options.SharedKey
+	state.sharedCipher, _ = cryptutil.NewAEADCipherFromBase64(cfg.Options.SharedKey)
+	state.cookieSecret, _ = base64.StdEncoding.DecodeString(cfg.Options.CookieSecret)
+
+	// used to load and verify JWT tokens signed by the authenticate service
+	state.encoder, err = jws.NewHS256Signer([]byte(cfg.Options.SharedKey), cfg.Options.GetAuthenticateURL().Host)
+	if err != nil {
+		return nil, err
+	}
+
+	state.refreshCooldown = cfg.Options.RefreshCooldown
+	state.jwtClaimHeaders = cfg.Options.JWTClaimsHeaders
+
+	// errors checked in ValidateOptions
+	state.authorizeURL, _ = urlutil.DeepCopy(cfg.Options.AuthorizeURL)
+	state.authenticateURL, _ = urlutil.DeepCopy(cfg.Options.AuthenticateURL)
+	state.authenticateDashboardURL = state.authenticateURL.ResolveReference(&url.URL{Path: dashboardPath})
+	state.authenticateSigninURL = state.authenticateURL.ResolveReference(&url.URL{Path: signinURL})
+	state.authenticateSignoutURL = state.authenticateURL.ResolveReference(&url.URL{Path: signoutURL})
+	state.authenticateRefreshURL = state.authenticateURL.ResolveReference(&url.URL{Path: refreshURL})
+
+	state.sessionStore, err = cookie.NewStore(func() cookie.Options {
+		return cookie.Options{
+			Name:     cfg.Options.CookieName,
+			Domain:   cfg.Options.CookieDomain,
+			Secure:   cfg.Options.CookieSecure,
+			HTTPOnly: cfg.Options.CookieHTTPOnly,
+			Expire:   cfg.Options.CookieExpire,
+		}
+	}, state.encoder)
+	if err != nil {
+		return nil, err
+	}
+	state.sessionLoaders = []sessions.SessionLoader{
+		state.sessionStore,
+		header.NewStore(state.encoder, httputil.AuthorizationTypePomerium),
+		queryparam.NewStore(state.encoder, "pomerium_session")}
+
+	authzConn, err := grpc.GetGRPCClientConn("authorize", &grpc.Options{
+		Addr:                    state.authorizeURL,
+		OverrideCertificateName: cfg.Options.OverrideCertificateName,
+		CA:                      cfg.Options.CA,
+		CAFile:                  cfg.Options.CAFile,
+		RequestTimeout:          cfg.Options.GRPCClientTimeout,
+		ClientDNSRoundRobin:     cfg.Options.GRPCClientDNSRoundRobin,
+		WithInsecure:            cfg.Options.GRPCInsecure,
+		ServiceName:             cfg.Options.Services,
+	})
+	if err != nil {
+		return nil, err
+	}
+	state.authzClient = envoy_service_auth_v2.NewAuthorizationClient(authzConn)
+
+	return state, nil
+}
+
+type atomicProxyState struct {
+	value atomic.Value
+}
+
+func newAtomicProxyState(state *proxyState) *atomicProxyState {
+	aps := new(atomicProxyState)
+	aps.Store(state)
+	return aps
+}
+
+func (aps *atomicProxyState) Load() *proxyState {
```

Comment:
How big of a performance hit is Load? Should we add a benchmark just to get a sense?

Reply:
Load is very lightweight. It consists of two pointer loads from memory, so almost no different than a normal pointer. Store uses the `cmpxchg` processor instruction.

Because we always do a full replace of the struct, we don't have to worry about modifying memory used by multiple threads, so we can avoid a mutex.